Fix warnings when compiling with AppleClang

# Bug reports:

AppleClang appears to emit more or different warnings than gcc and clang in our container. We need to fix them before we can test with `-Werror` on CI.

```
/Users/runner/runners/2.165.2/work/spectre/spectre/tests/Unit/DataStructures/Test_IndexIterator.cpp:29:44: warning: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Wsign-compare]
    CHECK(index_iterator.collapsed_index() == call_num);
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~
/usr/local/include/catch2/catch.hpp:17329:97: note: expanded from macro 'CHECK'
#define CHECK( ... ) INTERNAL_CATCH_TEST( "CHECK", Catch::ResultDisposition::ContinueOnFailure, __VA_ARGS__ )
                                                                                                ^~~~~~~~~~~
/usr/local/include/catch2/catch.hpp:2677:40: note: expanded from macro 'INTERNAL_CATCH_TEST'
        CATCH_INTERNAL_IGNORE_BUT_WARN(__VA_ARGS__); \
                                       ^~~~~~~~~~~
/usr/local/include/catch2/catch.hpp:154:76: note: expanded from macro 'CATCH_INTERNAL_IGNORE_BUT_WARN'
#    define CATCH_INTERNAL_IGNORE_BUT_WARN(...) (void)__builtin_constant_p(__VA_ARGS__)
                                                                           ^~~~~~~~~~~
```

--- a/tests/Unit/DataStructures/Test_IndexIterator.cpp
+++ b/tests/Unit/DataStructures/Test_IndexIterator.cpp
@@ -26,7 +26,7 @@ SPECTRE_TEST_CASE("Unit.DataStructures.IndexIterator",
     CHECK(index_iterator() == expected);
     CHECK(*index_iterator == expected);
     CHECK(index_iterator->indices() == expected.indices());
-    CHECK(index_iterator.collapsed_index() == call_num);
+    CHECK(index_iterator.collapsed_index() == static_cast<size_t>(call_num));
     ++index_iterator;
     ++call_num;
   };

--- a/tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp
+++ b/tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp
@@ -59,8 +59,9 @@ void test_swsh_coefficients_class_interface() noexcept {
         computed_coefficients.get_sharp_alm_info()->lmax);
   CHECK(precomputed_libsharp_lm.get_sharp_alm_info()->lmax ==
         expected_sharp_alm_info->lmax);
-  CHECK(precomputed_libsharp_lm.get_sharp_alm_info()->lmax ==
-        computed_coefficients.l_max());
+  CHECK(
+      static_cast<size_t>(precomputed_libsharp_lm.get_sharp_alm_info()->lmax) ==
+      computed_coefficients.l_max());
 
   CHECK(precomputed_libsharp_lm.get_sharp_alm_info()->nm ==
         computed_coefficients.get_sharp_alm_info()->nm);

--- a/tests/Unit/Utilities/Test_MakeArray.cpp
+++ b/tests/Unit/Utilities/Test_MakeArray.cpp
@@ -117,7 +117,7 @@ SPECTRE_TEST_CASE("Unit.Utilities.MakeArray", "[Unit][Utilities]") {
     }
     const auto array = make_array<MyNonCopyable<int>, 3>(std::move(vector));
     for (size_t i = 0; i < 3; ++i) {
-      CHECK(gsl::at(array, i).get() == i);
+      CHECK(gsl::at(array, i).get() == static_cast<int>(i));
     }
   }
 

